### ISSUE TITLE
samsite strict cause csrf error on callback

### ISSUE THREAD

**Issue by K7cl** [oauth2-proxy/oauth2-proxy#1663]
## Expected Behavior

login normually

## Current Behavior

When set cookie-samesite to "strict" will cause an 403 error on callback.
the csrf cookie will not include in request due to samesite policy. so that oauth2-proxy will not receive a valid csrf value and 403 triggered

## Possible Solution

Remove csrf validation on callback api (I think [state parameter defined by oauth](https://auth0.com/docs/secure/attack-protection/state-parameters) can prevent a csrf attack, I don't understand why we use a csrf cookie to repeat the work)
or give an option to set samesite for csrf cookie

## Steps to Reproduce (for bugs)

follow below cookie settings
cookie_name = "__Host-OAUTH"
cookie_secret = "**********"
cookie_expire = "24h"
cookie_secure = true
cookie_httponly = true
cookie_samesite = "strict"

## Context

I want make my application security, do not want use a workaround such as set global policy samesite to lax to avoid this issue.

## Your Environment

RHEL 8.5
porxy behind nginx
chrome 101.0.4951.67

- Version used: 7.2.1

**Comment by caizixian** [comment]
Some relevant links

https://github.com/oauth2-proxy/oauth2-proxy/issues/817#issuecomment-1176138162
https://stackoverflow.com/questions/59990864/what-is-difference-between-samesite-lax-and-samesite-strict
https://stackoverflow.com/questions/42216700/how-can-i-redirect-after-oauth2-with-samesite-strict-and-still-get-my-cookies

**Comment by K7cl** [comment]
> 

I use redis to store sessions, i think is good if state can also save into session when we enable it

**Comment by K7cl** [comment]
> The CSRF cookie is used to store the state before calling authentication server
> 
> ```
>     tenant.auth0.com/authorize?...&state=xyzABC123
> ```
> 
> when authentication server callbacks oauth2 proxy, then the state is retrieved back and it is validated against the CSRF cookie.
> 
> ```
>    /callback?...&state=xyzABC123
> ```
> 
> In conclusion, removing the CSRF cookie makes the state validation impossible.
> 
> CSRF cookie implementation seems to be the most simple way to store the state. As an alternative, the state could be stored in memory or in a external storage (e.g. local or session storage in the browser). Since cookies are send back and forth between the server and client on every HTTP request, a web storage solution would be most appropriate, since information would be only on the client side.

Oh, in this way, i think the best solution maybe is give an option or default to set samesite none for csrf cookie

**Comment by anetbnd** [comment]
Hello

I see the same issue in my setup (OIDC):

1.) I start the login from oauth2 proxy, which sets the CRSF token
2.) then I get redirected to the IDP to login
3.) after login and confirming, that I want to share my data with this app, the IDP is redirecting me to the callback page
4.) the callback page is not seeing the CRSF token anymore, because the redirect was initiated by the IDP
5.) now, if I restart the login, it is not stopping anymore at the IDP (because here I already logged in) and thus the redirect is coming initially from my app and it works without this error

If I set `cookie_samesite` to `lax` it works without issues also for the first login attempt.

However, it should also work with the flow above, and `strict`, since the OIDC flow described above is the normal way how to do it. So actually the callback page should always accept redirects started by the expected IDP.

I think I can workaround this issue, by adding javascript to the error-page. It does the following:
If the referer header is my expected IDP and if the Message is the expected CRSF error message, it will simply automatically change to the startpage of the app, which then works (because the second login attempt can read the CRSF token). 

Nevertheless, I think it would be better to do this kind of "restart the login" directly from go, when two conditions are fulfilled:
1.) the referer header must be the expected IDP
2.) the CRSF token cannot be read
In this case you can just show a HTML page, which says: 
"You are getting redirected to the application soon. If this does not work, please click here." 

Or is there another solution available?

**Comment by anetbnd** [comment]
I tested the javascript workaround now and it works. I added the following javascript to the error.html template:
```javascript
  {{ if .Message }}
  let errorMessage = "{{.Message}}";
  let allowedReferrers = [
          "https://<my-idp-host>/"
  ];
  if (errorMessage.includes('CSRF') && allowedReferrers.includes(document.referrer))
  {
    document.title = "Login Redirect";
    document.body = document.createElement("body");
    document.body.innerHTML = "<div class=\"content\">If the browser is not redirect automatically, please follow <a href='/'>this</a> link.</div>";
    window.location.href = "/";
  }
  {{ end }}
```

Obviously this is a very hacky solution. Is there any possibility to do this directly from oauth2 proxy?

**Comment by K7cl** [comment]
> 



> > > The CSRF cookie is used to store the state before calling authentication server
> > > ```
> > >     tenant.auth0.com/authorize?...&state=xyzABC123
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > when authentication server callbacks oauth2 proxy, then the state is retrieved back and it is validated against the CSRF cookie.
> > > ```
> > >    /callback?...&state=xyzABC123
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > In conclusion, removing the CSRF cookie makes the state validation impossible.
> > > CSRF cookie implementation seems to be the most simple way to store the state. As an alternative, the state could be stored in memory or in a external storage (e.g. local or session storage in the browser). Since cookies are send back and forth between the server and client on every HTTP request, a web storage solution would be most appropriate, since information would be only on the client side.
> > 
> > 
> > Oh, in this way, i think the best solution maybe is give an option or default to set samesite none for csrf cookie
> 
> It's probably not sufficient, because there is a chain of redirects from the identity provider, so when `/oauth2/callback` redirects to the main website content, the `_oauth2_proxy` cookie might not be sent if it has `SameSite=Strict`, and users will be presented with a log in page again (a refresh will fix it though, but bad UX).

yes, you are right. i simulate by edit the http response (set cookie samesite None at csrf and keep strict at auth cookie). the auth cookie can successfully get in callback request, but in next redirect to app home, the auth cookie will not included due to chain of redirect not meet the policy

**Comment by nicolo-danzi** [comment]
re-activate. Still got the problem

**Comment by K7cl** [comment]
After reading some article about how to use state parameter in Oauth2, I'm agree that we should use cookie to validate it.
[A blog](https://fusebit.io/blog/oauth-state-parameters-nodejs/#:~:text=Regular%20web%20app,Cookie%20or%20session)
[Auth0 official description](https://auth0.com/docs/secure/attack-protection/state-parameters#set-and-compare-state-parameter-values:~:text=You%20store%20something%20on%20the%20client%20application%20side%20(in%20cookies%2C%20session%2C%20or%20localstorage)%20that%20allows%20you%20to%20perform%20the%20validation.)

And also I investigate some other use case, but I'm surprised that most of them set cookies samesite None
Such as in microsoft account: https://login.live.com/
the state seems validate by cookie: Set-Cookie: OParams=xxxxxxxxxxx domain=login.live.com; Secure; path=/; SameSite=None; HttpOnly
also the cookie after login: Set-Cookie: __Host-MSAAUTHP=xxxxxxxxxx; expires=Thu, 11-Jan-2024 07:20:27 GMT; Secure; path=/; SameSite=None; HttpOnly
but the different is the callback page they use are not a 302 redirection. It's  a normal status 200 with html contain a js to do the redirection.

So, I think a solution may be similar as: https://github.com/oauth2-proxy/oauth2-proxy/issues/1663#issuecomment-1210712001
We can set CSRF cookies samesite None and define a static page at callback, use js code to redirect user to app home.

**Comment by miguelborges99** [comment]
What oauth2-proxy version are you using? Could you try with the latest version and enable the parameter described bellow, to check if you are really facing the issue that you are reporting:

--cookie-csrf-per-request=true

See [page](https://oauth2-proxy.github.io/oauth2-proxy/docs/configuration/overview/) for more details.

does it work?

If oauth2-proxy had a new parameter 

--cookie-csrf-samesite

with options: "lax", "strict", "none". By default, to avoid breaking the current behavior, if parameter is not defined then the csrf cookie should use the value used in --cookie-samesite.

Would this solve your issue?

**Comment by K7cl** [comment]
> What oauth2-proxy version are you using? Could you try with the latest version and enable the parameter described bellow, to check if you are really facing the issue that you are reporting:
> 
> --cookie-csrf-per-request=true
> 
> See the [page](https://oauth2-proxy.github.io/oauth2-proxy/docs/configuration/overview/) for more details.
> 
> Does it work?
> 
> If oauth2-proxy had the parameter
> 
> --cookie-csrf-samesite
> 
> with options: "lax", "strict", "none", would this solve your issue?
> 
> In case --cookie-csrf-samesite is not defined then csrf cookie should use the value in --cookie-samesite, to not break the current behavior.

I use latest version v7.4.0. `--cookie-csrf-samesite` won't help due to chain of redirect, see: https://github.com/oauth2-proxy/oauth2-proxy/issues/1663#issuecomment-1229110548

**Comment by miguelborges99** [comment]
Is CSRF cookie created in a domain which is not the one you expected? 
Maybe you have to whitelist the domains, using the option --whitelist-domain.

Or, the CSRF cookie does not exist at all, after login?
Are you able to see in your developer tools, where cookie got lost?

**Comment by K7cl** [comment]
the csrf cookie will not exist at callback. in developer tools you can see the cookie is refused to include in request due to the redirect chain not meet the samesite policy.

**Comment by miguelborges99** [comment]
I've created this [PR](https://github.com/oauth2-proxy/oauth2-proxy/pull/1947).
The PR allows you to define the samesite for the CSRF cookie with the option --cookie-csrf-samesite.
Set --cookie-csrf-samesite equal to one of the values you desire (e.g. --cookie-csrf-samesite=none).

Are you able to make a build, install it and check where the cookie goes?

**Comment by miguelborges99** [comment]
I'm quite confused by your explanation, do not take me wrong. 
If you set to none the samesite of CSRF cookie, and to strict the samesite of session cookie, as you intended in [this reply](https://github.com/oauth2-proxy/oauth2-proxy/issues/1663#issuecomment-1190159255) then I would expect that CSRF cookie would go through the chain.

OIDC works with redirection, the solution proposed by you breaks that. Maybe you want something that is very specific to your application. 

This is how far I can get. Sorry, for not being able to help you more.

**Comment by K7cl** [comment]
> I'm quite confused by your explanation, do not take me wrong. If you set to none the samesite of CSRF cookie, and to strict the samesite of session cookie, as you intended in [this reply](https://github.com/oauth2-proxy/oauth2-proxy/issues/1663#issuecomment-1190159255) then I would expect that CSRF cookie would go through the chain.
yes, in that case, CSRF cookie will involve in request but the session cookie will not. so everything will look normally on callback api, but at the next redirect to the app home page, the session cookie will not include, so that the user will be redirected back to login page. just like that reply said

> OIDC works with redirection, the solution proposed by you breaks that. Maybe you want something that is very specific to your application.
yeah, maybe it is. by my investigation, there are seldom app use samesite strict on their oauth cookie
> This is how far I can go. Sorry, for not being able to help you more.
> 
> Maybe a diagram could help to understand the flow that you have and where you have the issues, and why it does not work... Probably a diagram with your idea to solve the issue would also be fine.
now, let me repeat what my problem is, step by step. forget all above.

the expect is: 
use samesite strict at session cookie

the current situation:
samesite set to strict at config will got 403 error when callback api redirect from idp

the reason: 
samesite policy block csrf cookie send to server at callback, due to it was redirect from idp's domain

the probbly solution and test:
1, remove csrf cookie verify at callback api. NOT work, oauth are not able to verify state parameter
2, split set csrf cookie samesite to None. NOT work, csrf cookie will go but session cookie will get block at next redirect
3, split set csrf cookie samesite None and define a static page at callback, use js code to redirect user to app home not response header location parameter. YES, WORK. Test by [this reply](https://github.com/oauth2-proxy/oauth2-proxy/issues/1663#issuecomment-1210712001)

**Comment by miguelborges99** [comment]
I've changed the [pull request](https://github.com/oauth2-proxy/oauth2-proxy/pull/1947), to incorporate your solution. 

The PR allows you to define the samesite for the CSRF cookie with the option --cookie-csrf-samesite, and also allows you to have a static page at callback with the option --use-redirect-page. 

Set --cookie-csrf-samesite equal to one of the values you desire (e.g. --cookie-csrf-samesite=none). Possible values are "lax", "strict", "none".

To activate the static page at callback, set --use-redirect-page=true. This will call a redirect_page.html which will do a meta refresh redirect to yout redirect URL. 

You may also change the redirect_page.html template, by creating your own docker image with --custom-templates-dir option. Example:

- create folder for your image configuration
- create Dockerfile

Dockerfile
```
FROM quay.io/oauth2-proxy/oauth2-proxy:v7.4.0
USER 65532:65532

# Copy initialization scripts
COPY --chown=65532:root custom-templates/*.* ./custom-templates/
```

- create custom-templates folder and copy your own redirect_page.html into it.
- build docker image
- use in oauth2 configuration the following flag --custom-templates-dir=./custom-templates

**Comment by K7cl** [comment]
> I've changed the [pull request](https://github.com/oauth2-proxy/oauth2-proxy/pull/1947), to incorporate your solution.
> 
> The pull request allows you to define the samesite for the CSRF cookie with the option --cookie-csrf-samesite, and also allows you to have a static page at callback with the option --use-redirect-page.
> 
> Set --cookie-csrf-samesite equal to one of the values you desire (e.g. --cookie-csrf-samesite=none). Possible values are "lax", "strict", "none".
> 
> To activate the static page at callback, set --use-redirect-page=true. This will call a redirect_page.html which will perform a meta refresh redirect to the redirect URL.
> 
> Can you test this?
> 
> You are also able to change the redirect_page.html template, by creating your own docker image and using the --custom-templates-dir option. Example:
> 
> * create folder for your docker configuration
> * create a file named "Dockerfile" in this folder
> 
> Dockerfile
> 
> ```
> FROM quay.io/oauth2-proxy/oauth2-proxy:v7.4.0
> USER 65532:65532
> 
> # Copy initialization scripts
> COPY --chown=65532:root custom-templates/*.* ./custom-templates/
> ```
> 
> * inside the same folder, create a "custom-templates" folder and copy your own redirect_page.html into it.
> * build docker image
> * use in oauth2 configuration the following flag --custom-templates-dir=./custom-templates

Tested and FIXED the issues! Thanks!
Will this pr merge into main branch and include in next release?
config:
`cookie_expire = "24h"
cookie_secure = true
cookie_httponly = true
cookie_samesite = "strict"
cookie_csrf_samesite = "none"
use_redirect_page = true
custom_templates_dir = "dir-to-custom-templates"`

**Comment by miguelborges99** [comment]
Did you also tested without the custom_templates_dir, i.e. without overriding the redirect_page.html?
Did it worked?

I do not know when this will be included in the main branch, this is up to the mantainers.

**Comment by K7cl** [comment]
> Did you also tested without the custom_templates_dir, i.e. without overriding the redirect_page.html? Did it worked? Can you test it?
> 
> I do not know when this will be included in the main branch, this is up to the mantainers. The code also needs to be reviewed and there could be some disagreements on the existance of this "--use-redirect-page" option. Let's see,,,

Yes, I did. The first time I only put sign_in and error html in custom_templates_dir, and it works. after taht, I made my own redirect_page html and put in togather.

**Comment by miguelborges99** [comment]
Do not stale.

**Comment by miguelborges99** [comment]
Do not stale, reviewer is needed for #1947.

**Comment by CharlieC3** [comment]
Would like to see this resolved sooner than later, https://github.com/oauth2-proxy/oauth2-proxy/pull/1947 still needs to be reviewed.

**Comment by C-Duv** [comment]
#1947 still pending review

**Comment by jimethn** [comment]
Looks like #1947 got changes requested

**Comment by jimethn** [comment]
Still have this problem, does anyone know of a workaround?

**Comment by K7cl** [comment]
Do not stale, reviewer is needed for https://github.com/oauth2-proxy/oauth2-proxy/pull/1947.

**Comment by jeffmccune** [comment]
If you want this fixed and don't mind running an auth proxy compiled from a random dude on the internet, I patched tag v7.6.0 with the following one-liner.

The amd64 image is published at `quay.io/holos/oauth2-proxy:v7.6.0`

```patch
From 77a03ae2c309a4bd5ba6200f7432a8027b00c79f Mon Sep 17 00:00:00 2001
From: Jeff McCune <jeff@openinfrastructure.co>
Date: Mon, 6 May 2024 15:07:20 -0700
Subject: [PATCH] Force CSRF cookie samesite none

Fix for https://github.com/holos-run/holos/issues/164
---
 pkg/cookies/csrf.go | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/pkg/cookies/csrf.go b/pkg/cookies/csrf.go
index 9840b544..8c61f010 100644
--- a/pkg/cookies/csrf.go
+++ b/pkg/cookies/csrf.go
@@ -141,6 +141,9 @@ func (c *csrf) SetCookie(rw http.ResponseWriter, req *http.Request) (*http.Cooki
 		c.cookieOpts.CSRFExpire,
 		c.time.Now(),
 	)
+	// Always set the CSRF as None
+	cookie.SameSite = http.SameSiteNoneMode
+
 	http.SetCookie(rw, cookie)
 
 	return cookie, nil
-- 
2.30.2
```

**Comment by boolexp** [comment]
Is there any chance to have this issue fixed in the forthcoming release?

**Comment by nobletrout** [comment]
it took me a few hours, but I finally figured out the right thing to google and landed here. The cookie totally gets lost if you have a more complex bounce around from here to there and back again with the csrf cookie. Since the cookie policy applies both to login cookie and the csrf cookie, it would be pretty nice to have a separate csrf cookie samesite policy compared to the rest of cookie nonsense that happens with this.

**Comment by ahmad-hamade** [comment]
I came across the same issue. Is there a plan to include a fix in the upcoming release?

**Comment by marekvinkler** [comment]
My take on this is that having the same SameSite policy for the CSRF cookie and the session cookie is not a good idea. CSRF cookie must be at most lax, while the session cookie should be strict to prevent cross-site requests.

Moreover, when OAuth2 proxy is using Redis, what's the use for the cookie then?
